### PR TITLE
[Spec Decode] Lazy recovery in rejection sampling: compute recovered tokens in-kernel on first rejection

### PR DESCRIPTION
**Pull request by masterFoad**
## Summary

This PR changes rejection-sampling recovery from eager to lazy.

Today, the rejection sampler prepares recovery sampling work before it knows whether recovery is needed. In speculative decoding, each request needs at most one recovered token for a speculative step: the token at the first rejected draft position. If all draft tokens are accepted, recovery is not needed. If an early draft token is rejected, later draft positions are discarded, so recovery work for those later positions is wasted.

This PR moves recovery sampling into `rejection_random_sample_kernel`. The kernel first runs accept/reject logic, and only on the first rejection does it sample a recovered token from the residual distribution `(target_probs - draft_probs)+`. For `draft_probs=None`, it samples from the target distribution with the rejected draft token excluded, matching the existing no-draft-probs behavior.

The result is the same sampling distribution, but without eager full-vocab recovery buffers in the production path.

## What changed

Changed files:

```text
vllm/v1/sample/rejection_sampler.py
tests/v1/sample/test_rejection_sampler.py
```

Implementation changes:

- `rejection_random_sample_kernel` now computes recovery lazily on first rejection.
- Recovery uses the same exponential-race formulation as the eager helper: `score = residual_prob / Exp(1)`, and the argmax is the recovered token.
- `generate_recovery_seeds` creates one int64 seed per speculative token. The full-vocab random values are generated in-kernel only when recovery is actually needed.
- `USE_FP64_GUMBEL` is supported in the lazy path with a constexpr branch, so fp32 and fp64 paths specialize independently.
- The eager `sample_recovered_tokens` helper remains for tests/compatibility, but production rejection sampling no longer materializes recovered tokens eagerly.
- Greedy sampling remains on the existing greedy path.

## Why this helps

Let:

- `B` = batch size
- `k` = number of draft tokens
- `V` = vocab size
- `R` = number of requests in the batch that actually need recovery

The old eager recovery path prepares recovery candidates before knowing which draft positions will reject. The scoring work scales roughly like:

```text
B x k x V
```

Lazy recovery first determines whether a request rejects. It scans the vocab only for requests that actually need a recovered token:

```text
R x V, where R <= B
```

So even in the worst case, where every request rejects, lazy recovery removes the `k` multiplier:

```text
old worst case: B x k x V
new worst case: B x V
```

If all drafts are accepted, lazy recovery does zero full-vocab recovery work.

Example with `B=64`, `k=16`, and `V~128k`:

```text
old eager recovery scoring: 64 x 16 x 128k ~ 131M vocab lanes
new worst case: 64 x 128k ~ 8M vocab lanes
new all-accepted case: 0 recovery lanes
```

Approximate reduction by recovery incidence:

| Requests needing recovery | Rough recovery-work reduction |
|---:|---:|
| 100% | 16x |
| 25% | 64x |
| 10% | 160x |
| 1% | 1600x |
| 0% | eager still pays; lazy pays zero |

There is also a random-number-generation win. In the benchmarked `B=64`, `V=50,272`, `k=16` shape, the eager helper generated about:

```text
64 x 50,272 ~ 3.2M exponential values
```

The lazy production path uses:

```text
64 x 16 = 1,024 int64 recovery seeds
```

That is about `3,142x` fewer random values in that shape.

Memory-wise, the eager helper materializes a full-vocab recovery-noise buffer of shape `B x V`. The lazy production path eliminates that recovery buffer.

Measured eager helper allocation sizes:

| batch | vocab | dtype | eager recovery buffer | lazy production recovery buffer |
|---:|---:|---|---:|---:|
| 64 | 50,272 | fp32 | 12.27 MiB | 0 bytes |
| 64 | 50,272 | fp64 | 24.55 MiB | 0 bytes |
| 64 | 128,256 | fp32 | 32.00 MiB peak | 0 bytes |
| 64 | 128,256 | fp64 | 62.62 MiB | 0 bytes |

This is a recovery-buffer allocation measurement, not a claim that whole-server peak memory drops by exactly these amounts.

## Correctness

Random recovery is distribution-equivalent to the eager implementation, not bitwise-identical to the old RNG stream. Seeded outputs may differ from main, but remain deterministic within this implementation.

Correctness coverage added/verified:

- Direct eager-vs-lazy distribution comparison for both `draft_probs=None` and `draft_probs` present.
- Analytic residual-distribution tests.
- Large-vocab / multi-tile 128k behavior.
- `draft_probs=True` and `draft_probs=None` branches.
- fp64 Gumbel recovery.
- Zero residual-mass fallback.
- Seeded determinism/isolation.
- Reject-at-last-position.
- Mixed greedy/random batches.
- Greedy path smoke: baseline-vs-baseline and baseline-vs-patched outputs are bit-identical in an offline fixed-batch `temperature=0` check.

Focused GPU sampler suite result:

```text
84 passed, 17 warnings
```

### Mechanism-level benchmark

On A100, final-tree mechanism benchmarks show the lazy path faster in all 20 tested shapes:

```text
range: -10.00% to -36.21% median mechanism time
```

Representative rows:

| vocab | draft_probs | reject pos | baseline median | patched median | delta |
|---:|---|---|---:|---:|---:|
| 50,272 | false | none | 0.998496 ms | 0.654304 ms | -34.47% |
| 50,272 | false | 0 | 0.937088 ms | 0.791136 ms | -15.58% |
| 50,272 | true | 0 | 0.927824 ms | 0.835008 ms | -10.00% |
| 128,256 | false | none | 2.165376 ms | 1.392432 ms | -35.70% |
| 128,256 | false | 0 | 2.152240 ms | 1.731280 ms | -19.56% |
| 128,256 | true | 0 | 2.173248 ms | 1.832160 ms | -15.69% |

### End-to-end throughput

The previous preliminary `+9%` end-to-end throughput claim is removed. Final-tree end-to-end probes on small-model configurations were neutral/noisy, which is expected because the rejection sampler is only one part of total decode time.

Offline batch-64 continuity rerun:

| metric | baseline median | patched median | paired median delta | paired 95% CI |
|---|---:|---:|---:|---:|
| requests/s | 80.1982 | 79.5696 | +0.29% | [-4.16%, +7.43%] |
| output tok/s | 7699.0251 | 7638.6860 | +0.29% | [-4.16%, +7.43%] |
| total tok/s | 48760.4922 | 48378.3448 | +0.29% | [-4.16%, +7.43%] |

Note: the paired delta is the median of per-repetition paired deltas, not the ratio of the two medians.

Synthetic confirmatory runs also stayed scoped/noisy:

| config | result |
|---|---|
| k=8, synthetic acceptance 0.95 | +0.23% req/s, CI [-1.03%, +2.34%] |
| k=16, synthetic acceptance 0.50 | -0.75% req/s, CI [-1.36%, -0.21%], with acceptance length -1.80% |

The k=16/r=0.5 result is not treated as a mechanism regression: throughput is mediated by acceptance length, and the lower acceptance length more than explains the request-throughput delta. The mechanism-level benchmark remains the primary performance evidence for this PR.

## EAGLE / draft-probs coverage note

Validation showed default EAGLE3 in this tree uses greedy draft mode with `draft_logits=None`, which corresponds to the `draft_probs=None` path at rejection sampling time.

So the coverage statement is intentionally narrow:

- `temperature > 0` EAGLE traffic would exercise the same `NO_DRAFT_PROBS=True` lazy-recovery branch as ngram.
- It does not exercise the `draft_probs=True` residual-subtraction branch.
- The `draft_probs=True` branch is covered by unit tests, direct eager-vs-lazy equivalence tests, and mechanism benchmarks with `draft_probs=True`.
- This PR does not include a dedicated EAGLE end-to-end performance claim.

## Limitations / tradeoffs

- Random sampling is distribution-equivalent, not bitwise-RNG-identical to main.
- End-to-end throughput is expected to be near-neutral when the rejection sampler is a small share of total decode time.
- The performance claim is intentionally scoped to mechanism-level wins and recovery-buffer/RNG elimination, not a broad e2e speedup headline.
- Default EAGLE3 in this tree covers `draft_probs=None`; `draft_probs=True` is covered by tests and mechanism benchmarks rather than EAGLE e2e.

## Test plan

```bash
pytest tests/v1/sample/test_rejection_sampler.py -v
# 84 passed, 17 warnings
```

Additional validation performed:

- Offline greedy equivalence smoke with fixed batch and `temperature=0`.
- Temp=1 random/speculative path smoke confirming draft tokens are produced.
- A100 mechanism benchmark across vocab sizes, rejection positions, and `draft_probs` true/false.
- Recovery-buffer allocation probe for fp32/fp64 at 50k and 128k vocab.

## AI assistance disclosure

AI assistance was used to implement, test, benchmark, and draft this PR description. A human submitter reviewed the changes and scoped the claims to the evidence above.

### CODE REVIEW

**Comment by masterFoad** [comment]
In addition, the following results are after benchmark run on my laptop (WSL/ windows):

**Performance Gains (RTX 4060 Laptop GPU, 8GB VRAM):**
*   **Kernel Speedup:** **~8.0x** (0.27ms vs 2.15ms) on a targeted Triton kernel benchmark (8 positions, reject at pos 0).
*   **End-to-End Speculative Decoding Throughput:** **+14.7%** (15.6 tokens/s vs 13.6 tokens/s) using Qwen2.5-1.5B with Qwen2.5-0.5B drafter.

**Comment by masterFoad** [comment]
Update: I force-pushed a cleaned-up single-commit version of this PR.

Changes since the earlier review:
- Restored the thinking-budget logits-processing path so the PR stays scoped to speculative rejection sampling.
- Kept `output_token_ids` allocated before greedy/random branching, including the all-random path.
- Preserved `float64` uniform probabilities and `int32` output token IDs.
- Generate uniform probabilities for synthetic mode even when all requests are greedy.
- Kept `sample_recovered_tokens` as an eager compatibility/test helper, while the production rejection path now performs recovery lazily inside `rejection_random_sample_kernel`.

Local validation after cleanup:
- `py_compile` passed.
- `ruff-check` passed.
- `ruff-format` passed.
- `typos` passed.
- `pytest --collect-only tests/v1/sample/test_rejection_sampler.py` passed locally.

Remaining validation needed:
- I am working on tests and updated/verified benchmark.

The core change is now narrower: eager recovered-token sampling is removed from the production rejection path, but the eager helper remains for test/compatibility coverage.

**Comment by masterFoad** [comment]
Summary:
- Production lazy recovery now uses one inverse-CDF recovery uniform per draft token instead of full-vocab exponential recovery noise.
- This keeps the recovery distribution equivalent, but does not preserve bit-identical recovered token IDs versus the old exponential-race RNG stream.
- Added the benchmark and validation details to the PR body.

**Comment by masterFoad** [comment]
- A100 mechanism check passed. It shows the intended accepted-path optimization directly: eager recovery uses 3,217,408 full-vocab recovery random values per iteration in the artifact shape, while lazy uses 1,024 scalar recovery uniforms. That is 3142.0x fewer values and 1571.0x fewer recovery-buffer bytes.
- A100 all-accept mechanism timing was eager 1.284 ms vs lazy 0.140 ms, delta -89.10%.
- A100 first-reject tradeoff probe was slower for lazy in the artifact, eager 1.284 ms vs lazy 2.845 ms, so I documented this as a tradeoff instead of hiding it.
- A100 serving latency still shows workload-dependent behavior: standard ngram has +33.93% request/output throughput and strong p95/p99 tail wins, while synthetic high-accept has +4.14% throughput and E2EL p99 improvement but mixed TTFT/ITL tails.

**Comment by masterFoad** [comment]
I ran an additional A100 validation pass. In targeted kernel benchmarks, the lazy target-normalization path was **1.27×–2.8× faster** than materializing `target_probs` with softmax across the tested speculative-rejection shapes.

I also ran a small non-greedy ngram serving sanity check using an A/B/A/B setup: **2 baseline + 2 candidate runs**, **300 requests per run**, concurrency 4, fixed 128 output tokens, with the first 40 requests discarded as warmup. The candidate path was verified to execute. Averaged across the two baseline and two candidate runs, mean request latency was **1.2% lower** for the candidate. Speculative acceptance behavior was comparable: **84.69%** baseline vs **85.04%** candidate.
